habitat_sim/simulator.py

```diff
@@ -247,36 +256,96 @@ def initialize_agent(
         return agent
 
-    def get_sensor_observations(self) -> Dict[str, Union[ndarray, "Tensor"]]:
-        for _, sensor in self._sensors.items():
-            sensor.draw_observation()
+    @overload
+    def get_sensor_observations(
+        self, agent_ids: int = 0
+    ) -> Dict[str, Union[ndarray, "Tensor"]]:
+        ...
+
+    @overload
+    def get_sensor_observations(
+        self, agent_ids: List[int]
+    ) -> List[Dict[str, Union[ndarray, "Tensor"]]]:
+        ...
+
+    def get_sensor_observations(self, agent_ids: Union[int, List[int]] = 0):
+        return_single = False
+        if isinstance(agent_ids, int):
+            agent_ids = [agent_ids]
+            return_single = True
+
+        for agent_id in agent_ids:
+            agent_sensorsuite = self.__sensors[agent_id]
+            for _sensor_uuid, sensor in agent_sensorsuite.items():
+                sensor.draw_observation()
+
+        observations: List[Dict[str, Union[ndarray, "Tensor"]]] = []
+        for agent_id in agent_ids:
+            agent_observations: Dict[str, Union[ndarray, "Tensor"]] = {}
+            for sensor_uuid, sensor in self.__sensors[agent_id].items():
+                agent_observations[sensor_uuid] = sensor.get_observation()
+            observations.append(agent_observations)
+        if return_single:
+            return observations[0]
+        return observations
 
-        observations = {}
-        for sensor_uuid, sensor in self._sensors.items():
-            observations[sensor_uuid] = sensor.get_observation()
+    @property
+    def _default_agent(self) -> Agent:
+        # TODO Deprecate and remove
+        return self.get_agent(self._default_agent_id)
 
-        return observations
+    @property
+    def _sensors(self) -> Dict[str, "Sensor"]:
+        # TODO Deprecate and remove
+        return self.__sensors[self._default_agent_id]
 
-    def last_state(self):
+    def last_state(self) -> AgentState:
         return self._last_state
 
+    @overload
     def step(
-        self, action: Any, dt: float = 1.0 / 60.0
+        self, action: Union[str, int], dt: float = 1.0 / 60.0
     ) -> Dict[str, Union[bool, ndarray, "Tensor"]]:
+        ...
+
+    @overload
+    def step(
+        self, action: Dict[int, Union[str, int]], dt: float = 1.0 / 60.0
+    ) -> List[Dict[str, Union[bool, ndarray, "Tensor"]]]:
+        ...
+
+    def step(
+        self,
+        action: Union[str, int, Dict[int, Union[str, int]]],
+        dt: float = 1.0 / 60.0,
+    ) -> Union[
+        List[Dict[str, Union[bool, ndarray, "Tensor"]]],
+        Dict[str, Union[bool, ndarray, "Tensor"]],
+    ]:
         self._num_total_frames += 1
-        collided = self._default_agent.act(action)
-        self._last_state = self._default_agent.get_state()
+        return_single = False
+        if not isinstance(action, dict):
+            action = {self._default_agent_id: action}
+            return_single = True
+        agent_ids = sorted(action.keys())
```

Comment:
I really don't like this behavior, it's kind of wonky to return a List of Dicts in sorted order. Should really be a Dict of Dict's, right @erikwijmans ?

Reply:
Yeah, Dict of Dict's makes sense.  `get_sensor_observations` should also then be a dict of dicts for consistency.

Reply:
Less sure about this since the IDs are passed in directly as a list.

Reply:
Both solutions are annoying for different reasons. Iterating through a dict of a dicts is a bit painful and requires some subtle instancing to detect properly compared to a list at least but figuring out which ordering is return is a bit annoying.

We could always sort the list of agent dicts by agent_ids, but that can get a bit messy if not all agents generate observations...

Reply:
Yeah, I agree that it isn't ideal.  My thought is that even if dict of dicts is slightly more awkward, consistency is better because if one is list of dicts while the other is dict of dicts, that will require two different code paths.

What about returning an OrderedDict so `.values()` get's you the current list?